File: data_structures/ruby/binary_tree/binary_tree.rb
class BinaryTree
	attr_accessor :root, :left, :right

	def initialize(root_value, left = nil, right = nil)
		@root_value = root_value
		@left = left
		@right = right
	end

	def set_left_child(left_child_root, grandchild_left = nil, grandchild_right = nil)
		@left = BinaryTree.new(left_child_root, grandchild_left, grandchild_right)
	end

	def set_right_child(right_child_root, grandchild_left = nil, grandchild_right = nil)
		@right = BinaryTree.new(right_child_root, grandchild_left, grandchild_right)
	end

	def traverse(level=0)
		puts "Level #{level} Root: #{@root_value}"
		level += 1
		if left
			print "Left: " 
			@left.traverse(level)
		end
		if right
			print "Right: "
			@right.traverse(level)
		end
	end

end

# my_tree = BinaryTree.new(7)
# my_tree.set_left_child(4)
# my_tree.set_right_child(9)
# my_tree.left.set_left_child(13)
# my_tree.left.set_right_child(21)
# my_tree.right.set_right_child(101)
# my_tree.right.set_left_child(67)
# my_tree.traverse

File: data_structures/javascript/binary_tree/binary_tree.js
class BinaryTree {
  constructor(rootValue, left, right) {
    this.rootValue = rootValue;
    this.left = left || null;
    this.right = right || null;
  }

  setLeftChild(leftRoot, grandchildLeft, grandchildRight) {
    grandchildLeft = grandchildLeft || null;
    grandchildRight = grandchildRight || null;
    this.left = new BinaryTree(leftRoot, grandchildLeft, grandchildRight);
  }

  setRightChild(rightRoot, grandchildLeft, grandchildRight) {
    grandchildLeft = grandchildLeft || null;
    grandchildRight = grandchildRight || null;
    this.right = new BinaryTree(rightRoot, grandchildLeft, grandchildRight);
  }
}